flink-core/src/main/java/org/apache/flink/core/plugin/PluginUtils.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.plugin;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * Utility functions for the plugin mechanism.
+ */
+public final class PluginUtils {
+
+	private static final Logger log = LoggerFactory.getLogger(PluginUtils.class);
+
+	private PluginUtils() {
+		throw new AssertionError("Singleton class.");
+	}
+
+	public static void initPluginManagerSingletonFromRootFolder(@Nullable Path pluginsRootPath) {
+		Collection<PluginDescriptor> pluginDescriptorsForDirectory = Collections.emptyList();
+		if (pluginsRootPath != null) {
+			try {
+				pluginDescriptorsForDirectory =
+					new DirectoryBasedPluginDescriptorsFactory(pluginsRootPath).createPluginDescriptors();
+			} catch (IOException ioex) {
+				log.warn("Exception when locating plugins in root folder {}. Ignoring plugins.", ioex);
```

Comment:
same question, should we ignore the exceptions? If there is some missing plugin a job will fail later with a less readable error message. I would prefer to throw exceptions, but just make sure that code will never throw if `plugins` directory is empty.

Reply:
This is a bit more tricky questions. For example, what should happen to empty directories inside the `plugins` directory?

Reply:
Furthermore, are plugins somewhat optional or not? Should a failure from one plugin block the job execution or not?

Reply:
After offline discussion, making this strict and fast-failing as well.